src/conn/conn_compact.c

```diff
@@ -0,0 +1,202 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __compact_server_run_chk --
+ *     Check to decide if the compact server should continue running.
+ */
+static bool
+__compact_server_run_chk(WT_SESSION_IMPL *session)
+{
+    return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
+}
+
+/*
+ * __compact_server --
+ *     The compact server thread.
+ */
+static WT_THREAD_RET
+__compact_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *wt_session;
+    WT_SESSION_IMPL *session;
+    int exact;
+    const char *config, *key;
+    bool full_iteration;
+
+    session = arg;
+    conn = S2C(session);
+    wt_session = (WT_SESSION *)session;
+    cursor = NULL;
+    config = NULL;
+    key = NULL;
+    exact = 0;
+    full_iteration = false;
+
+    for (;;) {
+
+        /* When the entire metadata file has been parsed, take a break. */
+        if (full_iteration) {
+            full_iteration = false;
+            /*
+             * TODO: We may have processed just a few files, might as well sleep to give the system
+             * a break.
+             */
+            __wt_sleep(10, 0);
+        }
+
+        /* Only an explicit call to the compact API should signal the thread to wake up. */
+        if (!conn->background_compact.running) {
+            /*
+             * TODO: Depending on the previous state, we may not want to clear out the last key
+             * used. This could be useful if the server was paused to be resumed later.
+             */
+            key = NULL;
```

Comment:
Since we only consider keys with the `file:` prefix, maybe we should use that string as the initial key value. That way our first search wouldn't need to `cursor->next` past everything else in the metadata.

Reply:
Very good idea, thank you! I have made changes in ed8fc62b692e23a3d5817136458e718884a6f20f, let me know what you think.

Reply:
LGTM. I suggested this because for efficiency, but looking at your change it also makes the code easier to follow. A double win!